File: typescript/templates.ts
/**
 * Prompt 模板管理器
 * =================
 * 用法:
 *   import { templates } from './llm_sdk'
 *   templates.register('greet', { user: '用 {lang} 打招呼' })
 *   const reply = await templates.chatWithTemplate(client, 'greet', { lang: '日语' })
 *
 * 模板文件格式 (prompts/code_review.txt):
 *   system: 你是一个资深 {language} 工程师。
 *   ---
 *   审查以下代码，关注 {focus}:
 *
 *   {code}
 */

import * as fs from 'fs'
import * as path from 'path'
import { TemplateNotFoundError } from './errors'
import type { LLMClient, ChatOptions } from './client'

interface PromptTemplate {
  name: string
  userTemplate: string
  systemTemplate?: string
}

function renderTemplate(template: string, vars: Record<string, string>): string {
  return template.replace(/\{(\w+)\}/g, (_, key) => {
    if (!(key in vars)) throw new Error(`模板变量未提供: '${key}'`)
    return vars[key]
  })
}

function parseTemplateFile(name: string, content: string): PromptTemplate {
  content = content.trim()
  if (/^system:/i.test(content)) {
    const parts = content.split(/\n---\n/)
    if (parts.length >= 2) {
      const systemText = parts[0].replace(/^system:\s*/i, '').trim()
      const userText = parts.slice(1).join('\n---\n').trim()
      return { name, userTemplate: userText, systemTemplate: systemText }
    }
  }
  return { name, userTemplate: content }
}

export class TemplateRegistry {
  private registry = new Map<string, PromptTemplate>()

  register(name: string, opts: { user: string; system?: string }): void {
    this.registry.set(name, {
      name,
      userTemplate: opts.user.trim(),
      systemTemplate: opts.system?.trim(),
    })
  }

  loadDir(directory: string): number {
    const files = fs.readdirSync(directory).filter(f => f.endsWith('.txt'))
    for (const file of files) {
      const name = path.basename(file, '.txt')
      const content = fs.readFileSync(path.join(directory, file), 'utf-8')
      this.registry.set(name, parseTemplateFile(name, content))
    }
    return files.length
  }

  get(name: string): PromptTemplate {
    const tmpl = this.registry.get(name)
    if (!tmpl) throw new TemplateNotFoundError(name)
    return tmpl
  }

  list(): string[] {
    return [...this.registry.keys()]
  }

  async chatWithTemplate(
    llm: LLMClient,
    templateName: string,
    variables: Record<string, string>,
    opts: Omit<ChatOptions, 'system'> = {},
  ): Promise<string> {
    const tmpl = this.get(templateName)
    const userPrompt = renderTemplate(tmpl.userTemplate, variables)
    const system = tmpl.systemTemplate
      ? renderTemplate(tmpl.systemTemplate, variables)
      : undefined
    const result = await llm.chat(userPrompt, { ...opts, system })
    return typeof result === 'string' ? result : result.content
  }
}

export const templates = new TemplateRegistry()

File: go/templates.go
package llmclient

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// PromptTemplate holds a named prompt with optional system and user parts.
type PromptTemplate struct {
	Name           string
	UserTemplate   string
	SystemTemplate string
}

// Render replaces {key} placeholders with the provided variables.
func (t *PromptTemplate) Render(vars map[string]string) (user, system string, err error) {
	user, err = renderTemplate(t.UserTemplate, vars)
	if err != nil {
		return
	}
	if t.SystemTemplate != "" {
		system, err = renderTemplate(t.SystemTemplate, vars)
	}
	return
}

var placeholderRe = regexp.MustCompile(`\{(\w+)\}`)

func renderTemplate(tmpl string, vars map[string]string) (string, error) {
	var missing []string
	result := placeholderRe.ReplaceAllStringFunc(tmpl, func(match string) string {
		key := match[1 : len(match)-1]
		if val, ok := vars[key]; ok {
			return val
		}
		missing = append(missing, key)
		return match
	})
	if len(missing) > 0 {
		return "", fmt.Errorf("missing template variables: %v", missing)
	}
	return result, nil
}

// TemplateRegistry stores and retrieves PromptTemplates.
type TemplateRegistry struct {
	registry map[string]*PromptTemplate
}

// NewTemplateRegistry creates an empty registry.
func NewTemplateRegistry() *TemplateRegistry {
	return &TemplateRegistry{registry: make(map[string]*PromptTemplate)}
}

// Register adds a template by name.
func (r *TemplateRegistry) Register(name, userTmpl, systemTmpl string) {
	r.registry[name] = &PromptTemplate{
		Name:           name,
		UserTemplate:   strings.TrimSpace(userTmpl),
		SystemTemplate: strings.TrimSpace(systemTmpl),
	}
}

// LoadDir scans a directory for *.txt files and registers each as a template.
func (r *TemplateRegistry) LoadDir(dir string) (int, error) {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return 0, err
	}
	count := 0
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".txt") {
			continue
		}
		name := strings.TrimSuffix(e.Name(), ".txt")
		raw, err := os.ReadFile(filepath.Join(dir, e.Name()))
		if err != nil {
			return count, err
		}
		r.registry[name] = parseTemplateFile(name, string(raw))
		count++
	}
	return count, nil
}

// Get retrieves a template by name or returns TemplateNotFoundError.
func (r *TemplateRegistry) Get(name string) (*PromptTemplate, error) {
	t, ok := r.registry[name]
	if !ok {
		return nil, &TemplateNotFoundError{
			LLMError:     LLMError{Message: fmt.Sprintf("template not found: %q", name)},
			TemplateName: name,
		}
	}
	return t, nil
}

// List returns all registered template names.
func (r *TemplateRegistry) List() []string {
	names := make([]string, 0, len(r.registry))
	for k := range r.registry {
		names = append(names, k)
	}
	return names
}

// ChatWithTemplate renders the named template and calls the LLM.
func (r *TemplateRegistry) ChatWithTemplate(
	ctx context.Context,
	c *Client,
	templateName string,
	vars map[string]string,
	opts *ChatOpts,
) (string, error) {
	tmpl, err := r.Get(templateName)
	if err != nil {
		return "", err
	}
	userPrompt, systemPrompt, err := tmpl.Render(vars)
	if err != nil {
		return "", err
	}
	if opts == nil {
		opts = &ChatOpts{}
	}
	opts.System = systemPrompt
	return c.Chat(ctx, userPrompt, opts)
}

func parseTemplateFile(name, content string) *PromptTemplate {
	content = strings.TrimSpace(content)
	sysRe := regexp.MustCompile(`(?i)^system:\s*`)
	if sysRe.MatchString(content) {
		parts := strings.SplitN(content, "\n---\n", 2)
		if len(parts) == 2 {
			systemText := strings.TrimSpace(sysRe.ReplaceAllString(parts[0], ""))
			return &PromptTemplate{
				Name:           name,
				UserTemplate:   strings.TrimSpace(parts[1]),
				SystemTemplate: systemText,
			}
		}
	}
	return &PromptTemplate{Name: name, UserTemplate: content}
}
